src/DistributedComputations/distributed_architectures.jl

```diff
@@ -6,55 +6,109 @@ import Oceananigans.Architectures: device, cpu_architecture, arch_array, array_t
 import Oceananigans.Grids: zeros
 import Oceananigans.Utils: sync_device!, tupleit
 
+import Base
+
 #####
 ##### Partitioning
 #####
 
-# Form 3-tuple
-regularize_size(sz::Tuple{<:Int, <:Int}, Rx, Ry) = (sz[1], sz[2], 1)
-
-# Infer whether x- or y-dimension is indicated
-function regularize_size(N::Int, Rx, Ry)
-    if Rx == 1
-        return (1, N, 1)
-    elseif Ry == 1
-        return (N, 1, 1)
-    else
-        throw(ArgumentError("We can't interpret the 1D size $N for 2D partitions!"))
-    end
+struct Partition{Sx, Sy, Sz}
+    x :: Sx
+    y :: Sy
+    z :: Sz
 end
 
-regularize_size(sz::Array{<:Int, 1}) = regularize_size(sz[1])
-regularize_size(sz::Array{<:Int, 2}) = regularize_size((sz[1], sz[2]))
-regularize_size(sz::Tuple{<:Int})    = regularize_size(sz[1])
+"""
+    Partition(; x = 1, y = 1, z = 1)
+
+Return `Partition` representing the division of a domain in
+the `x` (first), `y` (second) and `z` (third) dimension
+
+Keyword arguments: 
+==================
+
+- `x`: partitioning of the first dimension 
+- `y`: partitioning of the second dimension
+- `z`: partitioning of the third dimension
+
+if supplied as positional arguments `x` will be the first argument, 
+`y` the second and `z` the third
+
+`x`, `y` and `z` can be `Int`, `Equal`, `Fractional` or `Sizes` 
+(see below)
+
+Examples:
+========
+
+```jldoctest
+julia> using Oceananigans; using Oceananigans.DistributedComputations
+
+julia> Partition(1, 4)
+Domain partitioning with (1, 4, 1) ranks
+├── x-partitioning: none
+├── y-partitioning: 4
+└── z-partitioning: none
+
+julia> Partition(x = Fractional(1, 2, 3, 4))
+Domain partitioning with (4, 1, 1) ranks
+├── x-partitioning: domain fractions: (0.1, 0.2, 0.3, 0.4)
+├── y-partitioning: none
+└── z-partitioning: none
+
+```
+"""
+Partition(x)       = Partition(validate_partition(x, 1, 1)...)
+Partition(x, y)    = Partition(validate_partition(x, y, 1)...)
+
+Partition(; x = 1, y = 1, z = 1) = Partition(validate_partition(x, y, z)...)
+
+Base.show(io::IO, p::Partition) =
+    print(io,
+    "Domain partitioning with $(ranks(p)) ranks", "\n",
+    "├── x-partitioning: $(ranks(p.x) == 1 ? "none" : p.x)", "\n",
```

Comment:
Should we say "1" rather than "none"?

The purpose of a display is to give as much code-relevant information about the content of a type as possible. So I think "1" is more accurate, while "none" is not right, it doesn't correspond to any julia type (unless we are using "nothing"), but in that case we should say "nothing" not "none").

Reply:
I like to differentiate from 1 because with 1 rank there is no partitioning. `nothing` is a good option

Reply:
probably it is just better to omit the directions in which rank == 1

Reply:
what's the difference between "1" and "nothing"? If you write `x=1`, isn't that the same thing as no partition?

Reply:
It's important that it matches the underlying code, if `partition.x` should match what we claim the `x` partitioning is

Reply:
ok let's go with nothing then

Reply:
ok, I think we need to change `validate_partition` as well in that case